### ISSUE TITLE
Gitlab unstable after upgrade (14.8 to 14.10)

### ISSUE THREAD

**Issue by red-gecko27** [sameersbn/docker-gitlab#2549]
Hi,

When I merge old branch (previously created in v14.8) with the web interface (in v14.10) repository breaks.

When pull or clone, this error message is displayed:

```
git pull
fatal: commit-graph requires overflow generation data but has none
fatal: the remote end hung up unexpectedly
```

**Comment by guangee** [comment]
> Hi,
> 
> When I merge old branch (previously created in v14.8) with the web interface (in v14.10) repository breaks.
> 
> When pull or clone, this error message is displayed:
> 
> ```
> git pull
> fatal: commit-graph requires overflow generation data but has none
> fatal: the remote end hung up unexpectedly
> ```

i meet the same error,maybe it is a bug! i guess it can broken the repository

**Comment by f-io** [comment]
Temporary workaround, go to gitaly relative path: @hashed/...  and switch `git config core.commitGraph false`

**Comment by stefanfritsch** [comment]
I have the same issue. A user broke a repo by doing a merge. I'd recommend everyone who still can should not upgrade to 14.10 (as gitlab doesn't support downgrades). f-io's workaround helped in our case. Perhaps some more info for people not up to speed on gitlab's internals:

1. There should be a `/home/git/data/repositories` dir in your container (which should be on a persistent volume).
2. Most of your repos will be in a subdirectory: `@hashed/`
3. To get the hash google for an online sha256 calculator. Take the project id number (visible directly below the project name in the web interface, even for broken repos), hash it
4. Now descend the directories according to the hash (i.e. the first subdir is the first two characters of the hash, and so on
5. In the final dir is a bare git repo of your project. Edit the `config` file and add `commitGraph = false` to the `[core]` section. Use `apt update && apt install nano` to get an editor if necessary

**Comment by f-io** [comment]
> 3. To get the hash google for an online sha256 calculator. Take the project id number (visible directly below the project name in the web interface, even for broken repos), hash it

Or just: Menu -> Admin -> Projects -> select the broken one -> Project Info: Gitaly relative path: @hashed/...

**Comment by stefanfritsch** [comment]
Thanks for the help on finding the hash, that's a lot nicer.

> You don't need to edit the config file manually, git is available.

How? `git config` tells me `fatal: not in a git directory` in bare repos.

In any case, I found git config always to be less intuitive and more error prone than just editing the config directly.

**Comment by f-io** [comment]
`@hashed/84/3d/843d984ß32940ß3290ß490ß2394ekfdsklfn.git# set git config core.commitGraph false`

In my opinion, you should always avoid manually editing. If you have a typo in your command nothing happen, if you screw up you config file you can break a lot.

**Comment by paolomainardi** [comment]
Thanks for posting some workarounds, you saved me for a planned upgrade today.

Are there any updates about what is causing this issue ?

cc @sachilles

**Comment by paolomainardi** [comment]
@f-io @stefanfritsch can you please check which git version has been bundled with docker-gitlab 14.10 ? 

It seems that default version on gitlab has been changed: https://gitlab.com/gitlab-org/gitaly/-/commit/0cb70a8df7f282299ef06a76072d541ea349e8d0

And it's always reccomended to use the bundled version with Gitaly: https://docs.gitlab.com/ee/install/installation.html#git 

You should have at least the version `v2.35.1.gl1`

**Comment by paolomainardi** [comment]
Yes thay have a patched git version and docker-gitlab seems that is not following the reccomendation to use the gitlaly bundled one, we should start by changing the installer here.

**Comment by paolomainardi** [comment]
For instance, the bundled one has all those patches applied: https://gitlab.com/gitlab-org/gitaly/-/tree/master/_support/git-patches/v2.35.1.gl1

**Comment by stefanfritsch** [comment]
> `@hashed/84/3d/843d984ß32940ß3290ß490ß2394ekfdsklfn.git# git config core.commitGraph false`
> 
> In my opinion, you should always avoid manually editing. If you have a typo in your command nothing happen, if you screw up you config file you can break a lot.

As I said, that command doesn't work in bare repos:

```
/home/git/data/repositories/@hashed/f6/5c/f65ccfbfec288565c1d414275985547799fde0ed286c85a50bd0ec5faa01d1ac.git# git config core.commitGraph false
fatal: not in a git directory
```

I have no idea why it works for you but not me when we have identical setups (as they're containerized) - but I'm not surprised. `git` is one of the most finicky CLI programs I know. E.g. kubectl is also extremely complex and powerful but somehow it just works. The less I have to interact with git's cli the better.

**Comment by stefanfritsch** [comment]
> I think this one should address the bug: #2551 at least to have version parity with Gitlab and follow the right installation path

Do you know if that will fix existing problems with repos or if will we have to fix them manually?

**Comment by paolomainardi** [comment]
@stefanfritsch i don't know because i didn't upgrade it but the evidence tell me that it should fix everything

**Comment by stefanfritsch** [comment]
That's wonderful to hear. Thanks a lot

**Comment by sachilles** [comment]
The PR #2513 is merged. Can you confirm that the issues you observed are fixed?

**Comment by paolomainardi** [comment]
Can you please give us some feedback ?

**Comment by f-io** [comment]
I can confirm 
``` 
git --version 
git version 2.35.1.gl1
```

I was able to `git commit-graph write` without breaking a project, `git -C HASH.git fsck ` is still fine. 
Finally the object store is valid again, had some unreadable commits from object database for commit-graph that required a graph write.

Need to check a few other things too, but looks good so far.

**Comment by tDo** [comment]
Thanks for the workarounds and fast patch-PR and merge :) (And in general thanks for all the work
you guys are putting into this project, it's really awesome and helps a lot!)

Would it maybe be possible to also mark the latest image with a ```14.10.0-1``` tag?
Thanks in advanced

**Comment by f-io** [comment]
`docker pull sameersbn/gitlab:latest && docker tag sameersbn/gitlab:latest sameersbn/gitlab:14.10.0-1`

**Comment by ahills60** [comment]
@paolomainardi @sachilles Pleased to confirm that it works for me too. Repos that would fail to clone and push to are now working as they used to. Thanks for the fix @paolomainardi!

**Comment by gjrtimmer** [comment]
I can confirm that the #2551 fixes all the problems described in this issue. Currently you can fidx your environment by using the tag `latest` until the `14.10.0` has been overwritten in de container repository.

**Comment by sachilles** [comment]
I would like to do that, but I noticed inconsistencies that should be fixed before.

1. In the container there are three git binaries in at least two versions. On the one hand under `/usr/local/bin/git` in version `git version 2.35.1.gl1`, furthermore under `/bin/git` in version `git version 2.36.0` and as `/usr/bin/git` in version `git version 2.36.0`.

2. In the config file `/home/git/gitlab/config/gitlab.yml` I think at least in the 'advanced settings' (section '## Git settings') the `bin_path` for git has to be changed. Currently it refers to `/usr/bin/git` and therefore uses git version 2.36.0 (see point 1).

3. Installing the `git-core` package via the `Dockerfile` and uninstalling via the installation script is inconvenient. If the package `git-core` is no longer needed after the installation of gitaly, then the installation and uninstallation should be done via the installation script. E.g. via the `BUILD_DEPENDENCIES` in the installation script.

I think these issues need to be addressed before integration into a release. Are there people who can contribute to the work? (Or at least for testing?)

**Comment by paolomainardi** [comment]
I can contribute on this

**Comment by paolomainardi** [comment]
@sachilles started here: https://github.com/sameersbn/docker-gitlab/pull/2553 have you any idea who and when the other git binaries are installed ?

**Comment by sachilles** [comment]
No, I was really surprised by this finding.

**Comment by kkimurak** [comment]
note: `/bin` is symlink for `usr/bin`.

```sh
$ docker exec -it gitlab-main ls -lha /bin /sbin /lib*
lrwxrwxrwx. 1 root root 7 Apr 16 02:54 /bin -> usr/bin
lrwxrwxrwx. 1 root root 7 Apr 16 02:54 /lib -> usr/lib
lrwxrwxrwx. 1 root root 9 Apr 16 02:54 /lib64 -> usr/lib64
lrwxrwxrwx. 1 root root 8 Apr 16 02:54 /sbin -> usr/sbin
```

**Comment by paolomainardi** [comment]
Ok that’s why everything is working even tho gitlab.yml path is wrong.

**Comment by sachilles** [comment]
So, we fixed 

>     2. In the config file `/home/git/gitlab/config/gitlab.yml` I think at least in the 'advanced settings' (section '## Git settings') the `bin_path` for git has to be changed. Currently it refers to `/usr/bin/git` and therefore uses git in version 2.36.0 (see point 1).

and
 
>     3. Installing the `git-core` package via the `Dockerfile` and uninstalling via the installation script is inconvenient. If the package `git-core` is no longer needed after the installation of gitaly, then the installation and uninstallation should be done via the installation script. E.g. via the `BUILD_DEPENDENCIES` in the installation script.

However, 

>     1. In the container there are three git binaries in at least two versions. On the one hand under `/usr/local/bin/git` in version `git version 2.35.1.gl1`, furthermore under `/bin/git` in version `git version 2.36.0` and as `/usr/bin/git` in version `git version 2.36.0`.

is still unfixed.

However, since the next security release is more important in my opinion, this point will have to wait.

**Comment by stefanfritsch** [comment]
Seems to be fixed here as well

**Comment by sachilles** [comment]
I guess we can close this ticket now, isn't it?
